proposals/lip-0049.md

```diff
@@ -223,135 +223,171 @@ When processing a cross-chain message `ccm`, the functions below may be called b
 All cross-chain messages must have the correct format, which is checked by the following logic:
 
 ```python
-validateFormat(ccm):
-    if size(ccm) > MAX_CCM_SIZE bytes:
-        return False
-    if ccm does not follow crossChainMessageSchema:
-        return False
-    return True
+def validateFormat(ccm: CCM) -> None:
+    # Try to encode ccm.
+    # This raises an error if the ccm does not follow the correct schema.
+    ccmBytes = encode(crossChainMessageSchema, ccm)
+    # Check ccm size.
+    if ccmBytes > MAX_CCM_SIZE:
+        raise Exception(f"Cross-chain message size is larger than {MAX_CCM_SIZE}.")
 ```
 
 #### apply
 
 The `apply` function applies a cross-chain message, in the context of a cross-chain update `ccu`.
 
 ```python
-# The ccu including this message is passed as context to this function
-apply(ccu, ccm):
-    if terminatedStateAccount(ccm.sendingChainID) exists:
-        ccm is discarded and has no further effect
-
-    # Call the beforeApplyCCM functions from other modules
-    # in particular, this will assign the message fee to the ccu sender
-    for each module mdl for which beforeApplyCCM exists:
-        mdl.beforeApplyCCM(ccu, ccm)
-        if the previous call fails or returns False:
+def apply(ccu: CCU, ccm: CCM) -> None:
+    ccmID = SHA-256(encode(crossChainMessageSchema, ccm))
+    # Apply the message if the logic is supported
+    if (ccm.moduleID, ccm.crossChainCommandID) is supported: 
+        # Create a state snapshot.
+
+        snapshot = snapshot of the current state
+        try: 
+            # Call the beforeCrossChainCommandExecution functions from other modules
+            # For example, the Token module assigns the message fee to the CCU sender.
+            for each module mdl for which beforeCrossChainCommandExecution exists:
+                mdl.beforeCrossChainCommandExecution(ccu, ccm)
+            
+            # Execute the cross-chain command
+            crossChainCommand = cross-chain command associated with (ccm.moduleID, ccm.crossChainCommandID)
+            crossChainCommand.verify(ccm)
+            crossChainCommand.execute(ccm)
+
+            # Call the afterCrossChainCommandExecution functions from other modules
+            for each module mdl for which beforeCrossChainCommandExecution exists:
+                mdl.afterCrossChainCommandExecution(ccu, ccm)     
+
+            # Emit processing event. If an error occurs, this event emission is also undone.
+            emitEvent(
+                moduleID = MODULE_ID_INTEROPERABILITY,
+                typeID = TYPE_ID_CCM_PROCESSED,
+                data = { "ccmID": ccmID, "status": CCM_PROCESSED_STATUS_COMMAND_APPLIED},
+                topics = [ccm.sendingChainID, ccm.receivingChainID]
+            )
+        except:
+            revert state to snapshot 
```

Comment:
In the CCU LIP you mention that the chain termination is not reverted, but here the whole state is still reverted to the last snapshot.

Reply:
fixed after STM

Reply:
All clear now.